lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs

```diff
@@ -106,84 +110,103 @@ random
     -> NonEmpty TxOut
     -> UTxO
     -> ExceptT ErrCoinSelection m (CoinSelection, UTxO)
-random opt outs utxo = do
+random opt outs utxo0 = do
     let descending = NE.toList . NE.sortBy (flip $ comparing coin)
     randomMaybe <- lift $ runMaybeT $ foldM
-        (processTxOut opt)
-        (utxo, mempty)
+        (makeSelection opt)
+        (utxo0, [])
         (descending outs)
     case randomMaybe of
-        Just (utxo', res) ->
-            return (res, utxo')
+        Just (utxo1, res) -> do
+            (utxo2, selection) <-
+                lift $ foldM
+                (improveTxOut opt)
+                (utxo1, mempty)
+                (reverse res)
+            return (selection, utxo2)
```

Comment:
If we swap the arguments of `improveTxOut`, we can simplify this expression a bit and avoid the naming `0`, `1`, and `2`:

```hs
Just (utxo', res) -> 
    lift $ foldM 
        (improveTxOut opt)
        (mempty, utxo')
        (reverse res)
```